### PR TITLE
PHPUnit 9 - All modules in single PR

### PR DESCRIPTION
**Pull request by lbajsarowicz**
### Description (*)
Result of branch update.

### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/238
https://github.com/magento/partners-magento2b2b/pull/95

### Fixed Issues (if relevant)

1. https://github.com/magento/magento2/issues/27500

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz
Unfortunately, we only accept Pull Requests to the following branches: 2.2-develop, 2.3-develop, 2.4-develop.
Please create a new Pull Request and use one of those as a target.

Thank you.

File: app/code/Magento/Amqp/Test/Unit/Setup/ConfigOptionsListTest.php [hunk 1]
```diff
@@ -184,15 +185,15 @@ public function getCreateConfigDataProvider()
                     ConfigOptionsList::INPUT_KEY_QUEUE_AMQP_SSL_OPTIONS => '{"ssl_option":"test"}',
                 ],
                 ['queue' => ['amqp' => [
-                            'host' => 'host',
```
Comment (ihor-sviziev) [inline app/code/Magento/Amqp/Test/Unit/Setup/ConfigOptionsListTest.php:187]: Would be good to fix formatting here, but it's too minor

File: app/code/Magento/Analytics/Test/Unit/Model/Config/Backend/CollectionTimeTest.php [hunk 2]
```diff
@@ -80,8 +80,7 @@ public function testAfterSave()
      */
     public function testAfterSaveWrongValue()
     {
-        $this->expectException(\Magento\Framework\Exception\LocalizedException::class);
-
+        $this->expectException('Magento\Framework\Exception\LocalizedException');
```
Comment (ihor-sviziev) [inline app/code/Magento/Analytics/Test/Unit/Model/Config/Backend/CollectionTimeTest.php:83]: Would be good to revert such changes, for now found in one file such case.

**Comment by lbajsarowicz** [comment]
@slavvka and @ihor-sviziev I had to skip 2 tests in CatalogImportExport module, as someone who wrote that test modified both protected methods and properties. To be rewritten from scratch later.

**Comment by lbajsarowicz** [comment]
The modules that left are:
image

We still don't have changes from phpunit8 branch:
- Catalog
- ConfigurableProduct
- Customer
- Sales
- Theme

The tests completely fail for Weee module with weird outcome, that needs to be investigated.

**Comment by slavvka** [comment]
> @slavvka and @ihor-sviziev I had to skip 2 tests in CatalogImportExport module, as someone who wrote that test modified both protected methods and properties. To be rewritten from scratch later.

Makes sense but please create issuse for that

**Review by slavvka — CHANGES_REQUESTED** [review]
@lbajsarowicz Looks like you overwrote composer.json
```
16:01:17 + docker exec a36292ca6261d7a041e7bc7e19a023ef19875e927e8b2aa71abcc057d890fd11 /bin/bash -c 'sudo -E -u centos composer install -d /var/www/html --no-interaction; '
16:01:18 Loading composer repositories with package information
16:01:18 Installing dependencies (including require-dev) from lock file
16:01:18 Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. It is recommended that you run `composer update` or `composer update <package name>`.
16:01:18 Your requirements could not be resolved to an installable set of packages.
16:01:18 
16:01:18   Problem 1
16:01:18     - This package requires php ~7.1.3||~7.2.0||~7.3.0 but your PHP version (7.4.5) does not satisfy that requirement.
16:01:18   Problem 2
16:01:18     - Installation request for magento/composer 1.6.x-dev -> satisfiable by magento/composer[1.6.x-dev].
16:01:18     - magento/composer 1.6.x-dev requires php ~7.1.3||~7.2.0||~7.3.0 -> your PHP version (7.4.5) does not satisfy that requirement.
16:01:18   Problem 3
16:01:18     - Installation request for pelago/emogrifier v2.2.0 -> satisfiable by pelago/emogrifier[v2.2.0].
16:01:18     - pelago/emogrifier v2.2.0 requires php ^5.5.0 || ~7.0.0 || ~7.1.0 || ~7.2.0 || ~7.3.0 -> your PHP version (7.4.5) does not satisfy that requirement.
16:01:18   Problem 4
16:01:18     - Installation request for magento/magento2-functional-testing-framework dev-3.0.0-RC1 -> satisfiable by magento/magento2-functional-testing-framework[dev-3.0.0-RC1].
16:01:18     - magento/magento2-functional-testing-framework dev-3.0.0-RC1 requires php ~7.2.0||~7.3.0 -> your PHP version (7.4.5) does not satisfy that requirement.
16:01:18   Problem 5
16:01:18     - magento/magento2-functional-testing-framework dev-3.0.0-RC1 requires php ~7.2.0||~7.3.0 -> your PHP version (7.4.5) does not satisfy that requirement.
16:01:18     - codeception/codeception 2.4.5 requires facebook/webdriver >=1.1.3 <2.0 -> satisfiable by magento/magento2-functional-testing-framework[dev-3.0.0-RC1].
16:01:18     - Installation request for codeception/codeception 2.4.5 -> satisfiable by codeception/codeception[2.4.5].
16:01:18 
```

**Comment by slavvka** [comment]
Unit @ php74:
```
12:15:26 PHP Fatal error:  Declaration of Magento\Catalog\Test\Unit\Model\Product\CopierTest::setUp() must be compatible with PHPUnit\Framework\TestCase::setUp(): void in /var/www/html/app/code/Magento/Catalog/Test/Unit/Model/Product/CopierTest.php on line 72
```

**Comment by slavvka** [comment]
Static @ php74:
```
composer.lock file is not up to date (composer.json file was modified)
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'ab9aaa5f2c8e2a9b9c2cb3d969df9204'
+'a40525b41b0d0290f4178e2cf6ba5266'
```

**Comment by slavvka** [comment]
```
Passed: 182, Failed: 6, Incomplete: 0, Skipped: 0.
Data set: /var/www/html/app/code/Magento/MediaContent
PHP version ~7.1.3||~7.2.0||~7.3.0 in component magento/module-media-content is inconsistent with version ~7.3.0||~7.4.0 in root composer.json
Failed asserting that an array is empty.
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:342
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:184
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:97
/var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:99

Data set: /var/www/html/app/code/Magento/MediaContentApi
PHP version ~7.1.3||~7.2.0||~7.3.0 in component magento/module-media-content-api is inconsistent with version ~7.3.0||~7.4.0 in root composer.json
Failed asserting that an array is empty.
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:342
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:184
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:97
/var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:99

Data set: /var/www/html/app/code/Magento/MediaContentCatalog
PHP version ~7.1.3||~7.2.0||~7.3.0 in component magento/module-media-content-catalog is inconsistent with version ~7.3.0||~7.4.0 in root composer.json
Failed asserting that an array is empty.
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:342
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:184
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:97
/var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:99

Data set: /var/www/html/app/code/Magento/MediaContentCms
PHP version ~7.1.3||~7.2.0||~7.3.0 in component magento/module-media-content-cms is inconsistent with version ~7.3.0||~7.4.0 in root composer.json
Failed asserting that an array is empty.
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:342
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:184
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:97
/var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:99

Data set: /var/www/html/app/code/Magento/MediaGalleryCatalog
PHP version ~7.1.3||~7.2.0||~7.3.0 in component magento/module-media-gallery-catalog is inconsistent with version ~7.3.0||~7.4.0 in root composer.json
Failed asserting that an array is empty.
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:342
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:184
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:97
/var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:99

Data set: /var/www/html
Command "validate" failed: ./composer.json is valid, but with a few warnings
<warning>See https://getcomposer.org/doc/04-schema.md for details on the schema</warning>
The lock file is not up to date with the latest changes in composer.json, it is recommended that you run `composer update` or `composer update <package name>`.
<warning>Defining autoload.psr-0 with an empty namespace prefix is a bad idea for performance</warning>
<warning>require.braintree/braintree_php : exact version constraints (3.35.0) should be avoided if the package follows semantic versioning</warning>
<warning>require.colinmollenhour/cache-backend-redis : exact version constraints (1.11.0) should be avoided if the package follows semantic versioning</warning>
<warning>require.colinmollenhour/credis : exact version constraints (1.11.1) should be avoided if the package follows semantic versioning</warning>
<warning>require.magento/magento-composer-installer : unbound version constraints (>=0.1.11) should be avoided</warning>
<warning>require.phpseclib/mcrypt_compat : exact version constraints (1.0.8) should be avoided if the package follows semantic versioning</warning>
<warning>require.tubalmartin/cssmin : exact version constraints (4.1.1) should be avoided if the package follows semantic versioning</warning>

/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:142
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:93
/var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:99
```

**Comment by slavvka** [comment]
```
Redundant dependencies found!

Module Magento\Tinymce3: hard [Magento\Cms]
```

**Comment by slavvka** [comment]
Static, 73, 74, EE, B2B:
```
Passed: 15390, Failed: 1, Incomplete: 0, Skipped: 0.
Data set: app/code/Magento/VersionsCms/Observer/AffectCmsPageRender.php
Module Magento\VersionsCms has external dependencies: hard [Magento\CMS]
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/DependencyTest.php:542
/var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
/var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/DependencyTest.php:545
```

**Comment by slavvka** [comment]
Static 73, 74, B2B:
```
PHP Code Sniffer detected 2 violation(s): 

FILE: ...p/code/Magento/AdvancedSearch/Test/Unit/Block/SearchDataTest.php
----------------------------------------------------------------------
FOUND 0 ERRORS AND 3 WARNINGS AFFECTING 3 LINES
----------------------------------------------------------------------
 73 | WARNING | [x] Multi-line function call not indented correctly;
    |         |     expected 16 spaces but found 12
 79 | WARNING | [x] Multi-line function call not indented correctly;
    |         |     expected 16 spaces but found 12
 80 | WARNING | [x] Multi-line function call not indented correctly;
    |         |     expected 12 spaces but found 8
----------------------------------------------------------------------
PHPCBF CAN FIX THE 3 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------



Failed asserting that 2 matches expected 0.
```

**Comment by lbajsarowicz** [comment]
> Unit @ php74 and @ php73 CE, EE, B2B:
> 
> ```
> 12:15:26 PHP Fatal error:  Declaration of Magento\Catalog\Test\Unit\Model\Product\CopierTest::setUp() must be compatible with PHPUnit\Framework\TestCase::setUp(): void in /var/www/html/app/code/Magento/Catalog/Test/Unit/Model/Product/CopierTest.php on line 72
> ```

We cannot check Module Catalog, as it's still waiting for merge to `phpunit8`

> Static, php73, 74, CE, EE, B2B
> 
> ```
> Passed: 182, Failed: 6, Incomplete: 0, Skipped: 0.
> Data set: /var/www/html/app/code/Magento/MediaContent
> PHP version ~7.1.3||~7.2.0||~7.3.0 in component magento/module-media-content is inconsistent with version ~7.3.0||~7.4.0 in root composer.json
> Failed asserting that an array is empty.
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:342
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:184
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:97
> /var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:99
> 
> Data set: /var/www/html/app/code/Magento/MediaContentApi
> PHP version ~7.1.3||~7.2.0||~7.3.0 in component magento/module-media-content-api is inconsistent with version ~7.3.0||~7.4.0 in root composer.json
> Failed asserting that an array is empty.
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:342
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:184
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:97
> /var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:99
> 
> Data set: /var/www/html/app/code/Magento/MediaContentCatalog
> PHP version ~7.1.3||~7.2.0||~7.3.0 in component magento/module-media-content-catalog is inconsistent with version ~7.3.0||~7.4.0 in root composer.json
> Failed asserting that an array is empty.
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:342
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:184
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:97
> /var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:99
> 
> Data set: /var/www/html/app/code/Magento/MediaContentCms
> PHP version ~7.1.3||~7.2.0||~7.3.0 in component magento/module-media-content-cms is inconsistent with version ~7.3.0||~7.4.0 in root composer.json
> Failed asserting that an array is empty.
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:342
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:184
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:97
> /var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:99
> 
> Data set: /var/www/html/app/code/Magento/MediaGalleryCatalog
> PHP version ~7.1.3||~7.2.0||~7.3.0 in component magento/module-media-gallery-catalog is inconsistent with version ~7.3.0||~7.4.0 in root composer.json
> Failed asserting that an array is empty.
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:342
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:184
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:97
> /var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:99
> 
> Data set: /var/www/html
> Command "validate" failed: ./composer.json is valid, but with a few warnings
> <warning>See https://getcomposer.org/doc/04-schema.md for details on the schema</warning>
> The lock file is not up to date with the latest changes in composer.json, it is recommended that you run `composer update` or `composer update <package name>`.
> <warning>Defining autoload.psr-0 with an empty namespace prefix is a bad idea for performance</warning>
> <warning>require.braintree/braintree_php : exact version constraints (3.35.0) should be avoided if the package follows semantic versioning</warning>
> <warning>require.colinmollenhour/cache-backend-redis : exact version constraints (1.11.0) should be avoided if the package follows semantic versioning</warning>
> <warning>require.colinmollenhour/credis : exact version constraints (1.11.1) should be avoided if the package follows semantic versioning</warning>
> <warning>require.magento/magento-composer-installer : unbound version constraints (>=0.1.11) should be avoided</warning>
> <warning>require.phpseclib/mcrypt_compat : exact version constraints (1.0.8) should be avoided if the package follows semantic versioning</warning>
> <warning>require.tubalmartin/cssmin : exact version constraints (4.1.1) should be avoided if the package follows semantic versioning</warning>
> 
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:142
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:93
> /var/www/html/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
> /var/www/html/dev/tests/static/testsuite/Magento/Test/Integrity/ComposerTest.php:99
> ```

@sivaschenko Could you take a look, as you maintain the `Media` modules.

**Comment by lbajsarowicz** [comment]
@ihor-sviziev / @slavvka - In Payment Method there was a change to PHPUnit between 8 and 9, for `assertLessThan` / `assertGreaterThan` from ($actual, $expected) to ($expected, $actual). That's why I replaced `assertLessThan` with `assertGreaterThan` in this specific case.

**Comment by lbajsarowicz** [comment]
EAV model also had 3 tests using `assertAttributeEquals` and `getAttribute`, which were supposed to get access to the private methods and properties. 

Marked with:
```
        $this->markTestSkipped('Testing protected / private methods / properties');
```

**Comment by lbajsarowicz** [comment]
`\Magento\ImportExport\Test\Unit\Helper\ReportTest::testGetExecutionTime` marked as invalid due to:
```
1) Magento\ImportExport\Test\Unit\Helper\ReportTest::testGetExecutionTime
TypeError: DateTime::diff() expects parameter 2 to be bool, null given
```

`$this->markTestIncomplete('Invalid mocks used for DateTime object. Investigate later.');`

**Comment by lbajsarowicz** [comment]
Module Customer has interesting feature after turning on strict types:
image
I'm going to handle that right now, but later on the implementation has to be  changed, if people want to use it with strict types.

**Comment by slavvka** [comment]
@lbajsarowicz Please create issues for all you skip/remove since it may be vital to sustain those tests as they can cover important functionality.

**Comment by ihor-sviziev** [comment]
I reviewed quite a big part of this or and so far didn’t find any issues. One thing that I noticed - this pr contains changes from 2.4-develop branch , that are missing in 2.4-develop-php74.
@slavvka i think we can run the tests and merge this pr

**Review by ihor-sviziev — APPROVED** [review]

**Comment by engcom-Alfa** [comment]
QA not applicable

**Comment by slavvka** [comment]
@lbajsarowicz Please create issues for each skip you introduce and ensure its number mentioned in the skip message.

**Comment by slavvka** [comment]
@lbajsarowicz Please look at the allure report for Unit Tests CE
[allure-report-ce (1).zip](https://github.com/magento/magento2/files/4540585/allure-report-ce.1.zip)

**Comment by lbajsarowicz** [comment]
Turns out that I skipped `setup/` Unit Tests somehow. I'm going to provide them rather tomorrow.

**Comment by slavvka** [comment]
[allure-report-ce (2).zip](https://github.com/magento/magento2/files/4548494/allure-report-ce.2.zip)

**Comment by slavvka** [comment]
PHP 7.4 Unit
[allure-report-ce (3).zip](https://github.com/magento/magento2/files/4554692/allure-report-ce.3.zip)

**Pull request merged**